README.md
# NumberFormatter - A Sublime Text Plugin

This [Sublime Text](http://www.sublimetext.com/) plugin formats plain numbers by adding a thousands separator to them. It also unformats them in reverse. See the [Features](#features) section below for details.

## Installation

* **Package Control**
  1. [Install Package Control](https://packagecontrol.io/installation)
  1. [Bring up the Command Palette](https://sublime-text.readthedocs.io/en/stable/reference/command_palette.html#how-to-use-the-command-palette) and type "Package Control: Install Package"
  1. Type "NumberFormatter" and press <kbd>enter</kbd>
* **Directly**
  1. Locate the `Packages` folder in the [data directory](http://docs.sublimetext.info/en/sublime-text-3/basic_concepts.html#the-data-directory)
  1. Download the [latest version of NumberFormatter](https://github.com/ilyakam/NumberFormatter/releases/latest)
  1. Extract the archive into the `Packages` folder
* **Development**
  1. [Follow the instructions on the `CONTRIBUTING` guide](./CONTRIBUTING.md#getting-started)

## Usage

* Linux: <kbd>ctrl</kbd><kbd>shift</kbd><kbd>comma</kbd>
* macOS: <kbd>command</kbd><kbd>shift</kbd><kbd>comma</kbd>
* Windows: <kbd>ctrl</kbd><kbd>shift</kbd><kbd>comma</kbd>

## Features:

As long as the number is partially or fully selected, or while the cursor is on the number, this plugin will format or unformat it in the following way:

| Before      | After       |
| ----------- | ----------- |
| `12345`     | `12,345`    |
| `12,345`    | `12345`     |
| `12345.67`  | `12,345.67` |
| `12,345.67` | `12345.67`  |

This plugin supports multiple cursors and multiple selections. It can process both formatted and unformatted numbers in one go.

## Options

To change the settings, simply edit the `NumberFormatter.sublime-settings` file that's included with the plugin:

* `decimal_separator` - the character that separates between whole and decimal numbers
* `thousands_separator` - the character that groups numbers in the thousands, millions, billions, etc.
* `format_thousands` - whether or not to add a thousands separator to numbers between 1000 and 9999

## Warnings

* Numbers with preceding zeros are formatted anyway (`000123` » `000,123`)
* When the `thousands_separator` is set to an empty space (`" "`), the whole number must be selected in order to unformat it (`12 345` » `12345`)

## Changelog

See [CHANGELOG.md](./CHANGELOG.md)

## Contributing

See [CONTRIBUTING.md](./CONTRIBUTING.md)

## License

See [LICENSE.md](./LICENSE.md)

## Software Credits

The development of this software was made possible using the following components:

* [Expand Selection to Whitespace](https://github.com/bits/ExpandSelectionToWhitespace-SublimeText) by Paul Sarena
  Licensed under: MIT License


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2018-05-15
### Added:
- Installation instructions to the `README`
- Command to "Format Numbers" to the Command Palette

### Fixed:
- Ensure that the repository name is addressed in the `CONTRIBUTING` guide

## [1.0.3] - 2018-05-14
### Fixed:
- Ensure that updating the settings take immediate effect

## [1.0.2] - 2018-05-13
### Fixed:
- Ensure that large numbers do not change due to rounding when formatted
- Ensure that empty selected regions do not throw an error in the console
- Ensure that a useless decimal dot (e.g., `1234.`) doesn't count as a digit

## [1.0.1] - 2018-05-06
### Fixed:
- Ensure non-numeric strings are ignored without any errors

## [1.0.0] - 2018-04-29 
### Added:
- Ability to format numbers by adding thousands separator(s): `12345` » `12,345`
- Ability to format numbers by removing thousands separator(s): `12,345` » `12345`
- Ability to choose any character for the decimal and thousands separators
- Ability to skip formatting four-digit numbers between `1000` and `9999`


number_formatter.py
import re
import sublime
import sublime_plugin

class FormatNumberCommand(sublime_plugin.TextCommand):
  def expand_region_to_whitespace(self, region):
    # Consider only the lines containing the region
    line_region = self.view.line(region)

    if not line_region:
      return None

    line_string = self.view.substr(line_region)

    # Looks for a whitespace boundary at the beginning of the region or between
    # the beginning of the region and the beginning of the line it starts on
    region_begin = region.begin() - line_region.begin()

    space_boundary_begin = self.get_space_boundary_begin(line_string,
                                                         region_begin)

    # Looks for whitespace between end of the region
    # and the end of the line it ends on
    region_end = region.end() - line_region.begin()

    space_boundary_end = self.get_space_boundary_end(line_string, region_end)

    # Returns region expanded to whitespace boundaries
    expanded_region_begin = line_region.begin() + space_boundary_begin
    expanded_region_end = line_region.begin() + space_boundary_end

    return sublime.Region(expanded_region_begin, expanded_region_end)

  def get_space_boundary_begin(self, str, pos):
    """
    Returns position of the closest whitespace boundary between `pos`
    and the beginning of the string
    """

    match = re.search(r"\S+$", str[:pos], re.UNICODE)

    if match:
      return pos - (match.end() - match.start())

    # Sets `pos` as the boundary when `str` is not preceded by a whitespace
    return pos

  def get_space_boundary_end(self, str, pos):
    """
    Returns position of the closest whitespace boundary between `pos`
    and the end of the string
    """

    match = re.search(r"\s", str[pos:], re.UNICODE)

    if match:
      return pos + match.start()

    return len(str)

  def is_number(self, str):
    """
    Strips off all locale characters from string and tests whether it's a number
    """
    result = False
    str = (str
           .replace(self.decimal_separator, ".")
           .replace(self.thousands_separator, ""))

    try:
      val = float(str)
      result = True

    except:
      pass

    return result

  def run(self, edit):
    settings = sublime.load_settings("NumberFormatter.sublime-settings")

    self.decimal_separator = settings.get("decimal_separator", ".")
    self.format_thousands = settings.get("format_thousands", True)
    self.thousands_separator = settings.get("thousands_separator", ",")

    expanded_regions = []
    selected_regions = self.view.sel()

    # Expands selected regions to the nearest whitespace
    for region in selected_regions:
      expanded_region = self.expand_region_to_whitespace(region)

      if expanded_region is None:
        continue

      if self.is_number(self.view.substr(expanded_region)):
        expanded_regions.append(expanded_region)

    # Formats the numbers in the selected regions
    for region in reversed(expanded_regions):
      substr = self.view.substr(region)

      if self.thousands_separator in substr:
        # Removes formatting from the number
        new_string = substr.replace(self.thousands_separator, "")

      else:
        decimal_number = None

        # Adds formatting to the number
        if bool(re.search(re.escape(self.decimal_separator), substr)):
          # Splits the number into whole and decimal number strings
          whole_number, decimal_number = substr.split(self.decimal_separator)

        else:
          whole_number = substr

        # Skips formatting four-digit numbers when set
        if (not self.format_thousands) and (1000 <= int(whole_number) <= 9999):
          continue

        new_string = ""

        # Decorates whole number strings by adding a comma to every third
        # character starting from the end:
        for idx, character in enumerate(reversed(whole_number)):
          if idx % 3 == 0 and idx > 0:
            character = character + self.thousands_separator

          new_string = character + new_string

        # Recombines the whole and decimal number strings into a single number
        if decimal_number is not None:
          new_string += self.decimal_separator + decimal_number

      self.view.replace(edit, region, new_string)
